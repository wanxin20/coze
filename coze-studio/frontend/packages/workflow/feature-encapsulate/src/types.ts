/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type StandardNodeType } from '@coze-workflow/base/types';
import { type WorkflowMode } from '@coze-workflow/base/api';
import { type PluginContext } from '@flowgram-adapter/free-layout-editor';

import { type EncapsulateResult } from './encapsulate';

export interface EncapsulateGlobalState {
  spaceId: string;
  flowMode: WorkflowMode;
  projectId?: string;
  info: {
    name?: string;
  };
}

export interface NodeMeta {
  description: string;
  icon: string;
  subTitle: string;
  title: string;
}

export type GetGlobalStateOption = (
  context: PluginContext,
) => EncapsulateGlobalState;
export type GetNodeTemplateOption = (
  context: PluginContext,
) => (type: StandardNodeType) => NodeMeta | undefined;
export type OnEncapsulateOption = (
  result: EncapsulateResult,
  ctx: PluginContext,
) => Promise<void>;

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}
