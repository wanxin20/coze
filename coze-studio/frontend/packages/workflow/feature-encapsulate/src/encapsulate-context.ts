/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { type StandardNodeType } from '@coze-workflow/base/types';
import { WorkflowMode } from '@coze-workflow/base/api';
import {
  EntityManager,
  type PluginContext,
  PlaygroundConfigEntity,
} from '@flowgram-adapter/free-layout-editor';

import { checkEncapsulateGray } from './utils';
import {
  type NodeMeta,
  type GetGlobalStateOption,
  type GetNodeTemplateOption,
} from './types';

@injectable()
export class EncapsulateContext {
  @inject(EntityManager)
  protected readonly entityManager: EntityManager;

  @inject(PlaygroundConfigEntity)
  private playgroundConfigEntity: PlaygroundConfigEntity;

  private pluginContext: PluginContext;

  private getGlobalStateOption: GetGlobalStateOption = () => ({
    spaceId: '',
    flowMode: WorkflowMode.Workflow,
    info: {
      name: '',
    },
  });

  private getNodeMetaTemplateOption: GetNodeTemplateOption = () => () =>
    undefined;

  setGetGlobalState(getGlobalState: GetGlobalStateOption) {
    this.getGlobalStateOption = getGlobalState;
  }

  setGetNodeTemplate(getNodeTemplate: GetNodeTemplateOption) {
    this.getNodeMetaTemplateOption = getNodeTemplate;
  }

  getNodeTemplate(type: StandardNodeType): NodeMeta | undefined {
    return this.getNodeMetaTemplateOption(this.pluginContext)(type);
  }

  setPluginContext(context: PluginContext) {
    this.pluginContext = context;
  }

  private get globalState() {
    return this.getGlobalStateOption(this.pluginContext);
  }

  get spaceId() {
    return this.globalState?.spaceId;
  }

  get flowName() {
    return this.globalState?.info?.name;
  }

  get flowMode() {
    return this.globalState?.flowMode;
  }

  get isChatFlow() {
    return this.globalState?.flowMode === WorkflowMode.ChatFlow;
  }

  get enabled() {
    return checkEncapsulateGray() && !this.playgroundConfigEntity.readonly;
  }

  get projectId() {
    return this.globalState?.projectId;
  }
}
