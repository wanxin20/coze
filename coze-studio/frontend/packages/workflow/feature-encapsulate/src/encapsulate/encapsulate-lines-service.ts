/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { groupBy, uniq } from 'lodash-es';
import { inject, injectable } from 'inversify';
import { type WorkflowJSON } from '@coze-workflow/base';
import {
  type WorkflowLineEntity,
  WorkflowNodeLinesData,
  type WorkflowPortEntity,
  type WorkflowNodeEntity,
  WorkflowLinesManager,
  type WorkflowEdgeJSON,
  type WorkflowLinePortInfo,
} from '@flowgram-adapter/free-layout-editor';

import { isNodesInSubCanvas } from '../utils/subcanvas';
import { type ConnectPortsInfo } from './types';

@injectable()
export class EncapsulateLinesService {
  @inject(WorkflowLinesManager) declare linesManager: WorkflowLinesManager;

  /**
   * Get a valid encapsulation connection port
   * @Param nodes Encapsulated array of nodes
   * @throws Error input line does not conform to encapsulation rules
   * @Throws Error Output line does not conform to encapsulation rules
   * @Returns Encapsulated input and output port objects
   */
  getValidEncapsulateConnectPorts(
    nodes: WorkflowNodeEntity[],
  ): ConnectPortsInfo {
    const inputLines = this.getEncapsulateNodesInputLines(nodes);
    const outputLines = this.getEncapsulateNodesOutputLines(nodes);

    if (!this.validateEncapsulateLines(inputLines)) {
      throw new Error('输入线不符合封装规则');
    }

    if (!this.validateEncapsulateLines(outputLines)) {
      throw new Error('输出线不符合封装规则');
    }

    return {
      inputLines,
      outputLines,
      fromPorts: uniq(inputLines.map(_line => _line.fromPort)),
      toPorts: uniq(
        outputLines
          .map(_line => _line.toPort)
          .filter(Boolean) as WorkflowPortEntity[],
      ),
    };
  }

  /**
   * Connects the specified ports to the given node.
   *
   * @param portInfo - Information about the ports to be connected.
   * @param node - The node to which the ports will be connected.
   *
   * This method iterates over the `fromPorts` and `toPorts` arrays, creating lines
   * between each port and the specified node using the `linesManager.createLine` method.
   */

  connectPortsToNode(portInfo: ConnectPortsInfo, node: WorkflowNodeEntity) {
    const { fromPorts, toPorts } = portInfo;

    fromPorts.forEach(_fromPort => {
      this.linesManager.createLine({
        from: _fromPort.node.id,
        fromPort: _fromPort.portID,
        to: node.id,
      });
    });

    toPorts.forEach(_toPort => {
      this.linesManager.createLine({
        from: node.id,
        to: _toPort.node.id,
        toPort: _toPort.portID,
      });
    });
  }

  /**
   * Get all input lines in the package range
   * @Param nodes flowchart node array
   * @Returns array of input lines in the encapsulated range
   */
  getEncapsulateNodesInputLines(
    nodes: WorkflowNodeEntity[],
  ): WorkflowLineEntity[] {
    return uniq(
      nodes
        .map(_node => _node.getData(WorkflowNodeLinesData).inputLines)
        .flat(),
    ).filter(
      _line =>
        _line.from &&
        !nodes.includes(_line.from) &&
        (isNodesInSubCanvas(nodes) || !this.isSubCanvasLinkLine(_line)),
    );
  }

  /**
   * Get all output lines in the package range
   * @Param nodes flowchart node array
   * @Returns array of output lines in the encapsulated range
   */
  getEncapsulateNodesOutputLines(
    nodes: WorkflowNodeEntity[],
  ): WorkflowLineEntity[] {
    return uniq(
      nodes
        .map(_node => _node.getData(WorkflowNodeLinesData).outputLines)
        .flat(),
    ).filter(
      _line =>
        _line.to &&
        !nodes.includes(_line.to) &&
        (isNodesInSubCanvas(nodes) || !this.isSubCanvasLinkLine(_line)),
    );
  }

  /**
   * Verify that the input and output lines within the package range comply with the package rules
   */
  validateEncapsulateLines(lines: WorkflowLineEntity[]): boolean {
    const isFromPortUniq =
      uniq(lines.map(_line => _line.fromPort)).length === 1;
    const isToPortUniq = uniq(lines.map(_line => _line.toPort)).length === 1;
    return isFromPortUniq || isToPortUniq;
  }

  /**
   * Create encapsulation connection
   * @param ports
   * @param subFlowNode
   * @returns
   */
  createEncapsulateLines(
    ports: ConnectPortsInfo,
    subFlowNode: WorkflowNodeEntity,
  ) {
    const inputLines: WorkflowLineEntity[] = [];
    const outputLines: WorkflowLineEntity[] = [];
    ports.inputLines.forEach(line => {
      const inputLine = this.linesManager.createLine({
        from: line.from.id,
        fromPort: line.fromPort.portID,
        to: subFlowNode.id,
      });

      if (!inputLine) {
        throw new Error('create input line failed');
      }
      inputLines.push(inputLine);
    });

    ports.outputLines.forEach(line => {
      if (!line.to) {
        return;
      }

      const outputLine = this.linesManager.createLine({
        from: subFlowNode.id,
        to: line.to.id,
        toPort: line.toPort?.portID,
      });

      if (!outputLine) {
        throw new Error('create output line failed');
      }

      outputLines.push(outputLine);
    });

    return {
      inputLines,
      outputLines,
    };
  }

  /**
   * Create unsealing line
   * @param options
   */
  createDecapsulateLines(options: {
    node: WorkflowNodeEntity;
    workflowJSON: WorkflowJSON;
    startNodeId: string;
    endNodeId: string;
    idsMap: Map<string, string>;
  }) {
    const { node, startNodeId, endNodeId, idsMap, workflowJSON } = options;
    const edges = [
      ...workflowJSON.edges,
      // Lines in the sub-canvas
      ...workflowJSON.nodes
        .map(n => n.edges)
        .filter(Boolean)
        .flat(),
    ] as WorkflowEdgeJSON[];

    const edgesGroup = groupBy(edges, edge => {
      if (edge.sourceNodeID === startNodeId) {
        return 'input';
      }
      if (edge.targetNodeID === endNodeId) {
        return 'output';
      }

      return 'internal';
    });

    // interconnect
    const internalLines = this.createDecapsulateInternalLines(
      edgesGroup.internal || [],
      idsMap,
    );

    // input connection
    const inputLines = this.createDecapsulateInputLines(
      edgesGroup.input || [],
      node,
      idsMap,
    );

    // output connection
    const outputLines = this.createDecapsulateOutputLines(
      edgesGroup.output || [],
      node,
      idsMap,
    );

    return {
      inputLines,
      outputLines,
      internalLines,
    };
  }

  /**
   * Create an unblocked internal connection
   * @param internalEdges
   * @param idsMap
   * @returns
   */
  private createDecapsulateInternalLines(
    internalEdges: WorkflowEdgeJSON[],
    idsMap: Map<string, string>,
  ) {
    const createdLines: WorkflowLineEntity[] = [];
    internalEdges.forEach(edge => {
      if (!idsMap.has(edge.sourceNodeID) || !idsMap.has(edge.targetNodeID)) {
        return;
      }

      const line = {
        from: idsMap.get(edge.sourceNodeID) as string,
        to: idsMap.get(edge.targetNodeID) as string,
        fromPort: edge.sourcePortID,
        toPort: edge.targetPortID,
      };

      if (line.fromPort === 'loop-function-inline-output') {
        line.from = `LoopFunction_${line.from}`;
      }

      if (line.toPort === 'loop-function-inline-input') {
        line.to = `LoopFunction_${line.to}`;
      }

      if (line.fromPort === 'batch-function-inline-output') {
        line.from = `BatchFunction_${line.from}`;
      }

      if (line.toPort === 'batch-function-inline-input') {
        line.to = `BatchFunction_${line.to}`;
      }

      this.createLine(line, createdLines);
    });
    return createdLines;
  }

  /**
   * Create an unsealed input connection
   * @param inputEdges
   * @param node
   * @param idsMap
   * @returns
   */
  private createDecapsulateInputLines(
    inputEdges: WorkflowEdgeJSON[],
    node: WorkflowNodeEntity,
    idsMap: Map<string, string>,
  ) {
    const createdLines: WorkflowLineEntity[] = [];
    const { inputLines } = node.getData(WorkflowNodeLinesData);

    // Encapsulate multiple starts + encapsulate multiple upstream inputs, do not create connections
    if (inputLines.length > 1 && inputEdges.length > 1) {
      return createdLines;
    }

    inputLines.forEach(inputLine => {
      inputEdges.forEach(edge => {
        if (!idsMap.has(edge.targetNodeID)) {
          return;
        }

        this.createLine(
          {
            from: inputLine.from.id,
            fromPort: inputLine.fromPort.portID,
            to: idsMap.get(edge.targetNodeID) as string,
            toPort: edge.targetPortID,
          },
          createdLines,
        );
      });
    });

    return createdLines;
  }

  /**
   * Create an unsealed output connection
   * @param outputEdges
   * @param node
   * @param idsMap
   * @returns
   */
  private createDecapsulateOutputLines(
    outputEdges: WorkflowEdgeJSON[],
    node: WorkflowNodeEntity,
    idsMap: Map<string, string>,
  ) {
    const createdLines: WorkflowLineEntity[] = [];
    const { outputLines } = node.getData(WorkflowNodeLinesData);

    // Encapsulate multiple outputs + Encapsulate multiple downstream outputs without creating wires
    if (outputLines.length > 1 && outputEdges.length > 1) {
      return createdLines;
    }

    outputLines.forEach(outputLine => {
      outputEdges.forEach(edge => {
        if (!idsMap.has(edge.sourceNodeID)) {
          return;
        }

        this.createLine(
          {
            from: idsMap.get(edge.sourceNodeID) as string,
            fromPort: edge.sourcePortID,
            to: outputLine.to?.id,
            toPort: outputLine.toPort?.portID,
          },
          createdLines,
        );
      });
    });

    return createdLines;
  }

  /**
   * Create a connection
   * @param info
   * @param createdLines
   */
  private createLine(
    info: WorkflowLinePortInfo,
    createdLines?: WorkflowLineEntity[],
  ) {
    const line = this.linesManager.createLine(info);

    if (line && createdLines) {
      createdLines.push(line);
    }
  }

  private isSubCanvasLinkLine(line: WorkflowLineEntity): boolean {
    if (
      // The last wire in the loop
      line.toPort?.portID === 'loop-function-inline-input' ||
      // The first wire in the loop
      line.fromPort.portID === 'loop-function-inline-output'
    ) {
      return true;
    }
    if (
      line.toPort?.portID === 'batch-function-inline-input' ||
      line.fromPort.portID === 'batch-function-inline-output'
    ) {
      return true;
    }
    if (
      line.fromPort.portID === 'loop-output-to-function' &&
      line.toPort?.portID === 'loop-function-input'
    ) {
      return true;
    }
    if (
      line.fromPort.portID === 'batch-output-to-function' &&
      line.toPort?.portID === 'batch-function-input'
    ) {
      return true;
    }
    return false;
  }
}
