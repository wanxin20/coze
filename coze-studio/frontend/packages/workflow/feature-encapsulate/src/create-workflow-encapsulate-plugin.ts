/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  definePluginCreator,
  type PluginCreator,
} from '@flowgram-adapter/free-layout-editor';

import { WorkflowEncapsulateContainerModule } from './workflow-encapsulate-container-module';
import { EncapsulateValidatorsContainerModule } from './validators';
import { EncapsulateValidateManager } from './validate';
import {
  type GetGlobalStateOption,
  type GetNodeTemplateOption,
  type OnEncapsulateOption,
} from './types';
import { EncapsulateRenderContainerModule } from './render';
import { EncapsulateContext } from './encapsulate-context';
import { EncapsulateManager, EncapsulateService } from './encapsulate';

interface EncapsulatePluginOptions {
  getNodeTemplate?: GetNodeTemplateOption;
  getGlobalState?: GetGlobalStateOption;
  onEncapsulate?: OnEncapsulateOption;
}

export const createWorkflowEncapsulatePlugin: PluginCreator<EncapsulatePluginOptions> =
  definePluginCreator<EncapsulatePluginOptions>({
    onInit(ctx, options) {
      ctx.get<EncapsulateManager>(EncapsulateManager).init();

      ctx.get<EncapsulateContext>(EncapsulateContext).setPluginContext(ctx);

      if (options.getNodeTemplate) {
        ctx
          .get<EncapsulateContext>(EncapsulateContext)
          .setGetNodeTemplate(options.getNodeTemplate);
      }

      if (options.getGlobalState) {
        ctx
          .get<EncapsulateContext>(EncapsulateContext)
          .setGetGlobalState(options.getGlobalState);
      }

      if (options.onEncapsulate) {
        ctx.get<EncapsulateService>(EncapsulateService).onEncapsulate(res => {
          options?.onEncapsulate?.(res, ctx);
        });
      }
    },
    onDispose(ctx) {
      ctx.get<EncapsulateValidateManager>(EncapsulateValidateManager).dispose();
      ctx.get<EncapsulateService>(EncapsulateService).dispose();
      ctx.get<EncapsulateManager>(EncapsulateManager).dispose();
    },
    containerModules: [
      WorkflowEncapsulateContainerModule,
      EncapsulateRenderContainerModule,
      EncapsulateValidatorsContainerModule,
    ],
  });
