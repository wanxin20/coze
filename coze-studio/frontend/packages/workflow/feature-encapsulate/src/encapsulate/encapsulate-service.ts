/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { injectable, inject } from 'inversify';
import { WorkflowMode } from '@coze-workflow/base/api';
import {
  reporter,
  StandardNodeType,
  type WorkflowJSON,
} from '@coze-workflow/base';
import { I18n } from '@coze-arch/i18n';
import { Toast } from '@coze-arch/coze-design';
import { HistoryService } from '@flowgram-adapter/free-layout-editor';
import {
  WorkflowDocument,
  type WorkflowNodeEntity,
  WorkflowSelectService,
} from '@flowgram-adapter/free-layout-editor';
import { delay, Emitter, logger } from '@flowgram-adapter/common';

import { EncapsulateValidateService } from '../validate';
import { isNodeInSubCanvas } from '../utils/subcanvas';
import { randomNameSuffix } from '../utils/random-name-suffix';
import { hasSubCanvasNodes } from '../utils/has-sub-canvas-nodes';
import { getSubWorkflowInfo } from '../utils';
import { EncapsulateGenerateService } from '../generate';
import { EncapsulateContext } from '../encapsulate-context';
import { EncapsulateApiService } from '../api';
import {
  type EncapsulateErrorResult,
  type EncapsulateResult,
  type EncapsulateService,
} from './types';
import { EncapsulateVariableService } from './encapsulate-variable-service';
import { EncapsulateNodesService } from './encapsulate-nodes-service';
import { EncapsulateLinesService } from './encapsulate-lines-service';

@injectable()
export class EncapsulateServiceImpl implements EncapsulateService {
  private onEncapsulateEmitter = new Emitter<EncapsulateResult>();

  public readonly onEncapsulate = this.onEncapsulateEmitter.event;

  @inject(EncapsulateValidateService)
  private encapsulateValidateService: EncapsulateValidateService;

  @inject(EncapsulateGenerateService)
  private encapsulateGenerateService: EncapsulateGenerateService;

  @inject(WorkflowSelectService)
  private workflowSelectService: WorkflowSelectService;

  @inject(EncapsulateNodesService)
  private encapsulateNodesService: EncapsulateNodesService;

  @inject(EncapsulateApiService)
  private encapsulateApiService: EncapsulateApiService;

  @inject(EncapsulateContext)
  private encapsulateContext: EncapsulateContext;

  @inject(EncapsulateLinesService)
  private encapsulateLinesService: EncapsulateLinesService;

  @inject(EncapsulateVariableService)
  private encapsulateVariableService: EncapsulateVariableService;

  @inject(HistoryService)
  private historyService: HistoryService;

  @inject(WorkflowDocument)
  private workflowDocument: WorkflowDocument;

  private encapsulating = false;
  private decapsulating = false;

  validate() {
    const { selectedNodes } = this.workflowSelectService;
    return this.encapsulateValidateService.validate(selectedNodes);
  }

  canEncapsulate(): boolean {
    return this.encapsulateContext.enabled;
  }

  async encapsulate() {
    if (!this.canEncapsulate()) {
      return this.encapsulateError('encapsulate is not enabled');
    }

    if (this.encapsulating) {
      return this.encapsulateError('encapsulating');
    }

    const name = `${
      this.encapsulateContext.flowName || ''
    }_sub_${randomNameSuffix()}`;
    const { selectedNodes } = this.workflowSelectService;

    if (selectedNodes.length < 2) {
      return this.encapsulateError('at least 2 nodes');
    }

    reporter.event({ eventName: 'workflow_encapsulate' });

    this.encapsulating = true;

    let res;
    try {
      res = await this.encapsulateNodes(name, selectedNodes);
      // Encapsulation complete Select the encapsulated node
      if (res.success) {
        this.workflowSelectService.selectNode(res.subFlowNode);
      }

      this.onEncapsulateEmitter.fire(res);
    } finally {
      this.encapsulating = false;
    }
    return res;
  }

  canDecapsulate(node: WorkflowNodeEntity) {
    return (
      this.encapsulateContext.enabled &&
      node.flowNodeType === StandardNodeType.SubWorkflow
    );
  }

  async decapsulate(node: WorkflowNodeEntity) {
    if (!node || !this.canDecapsulate(node) || this.decapsulating) {
      return;
    }

    reporter.event({ eventName: 'workflow_decapsulate' });
    this.decapsulating = true;
    try {
      await this.decapsulateNode(node);
    } finally {
      this.decapsulating = false;
    }
  }

  async decapsulateNode(node: WorkflowNodeEntity) {
    const info = getSubWorkflowInfo(node);

    if (!info) {
      return;
    }

    // Get process data
    const workflow = await this.encapsulateApiService.getWorkflow(
      info.spaceId && info.spaceId !== '0'
        ? info.spaceId
        : this.encapsulateContext.spaceId,
      info.workflowId,
      info.workflowVersion,
    );

    if (!workflow) {
      return;
    }

    if (workflow.nodes.length <= 2) {
      return;
    }

    // If you unseal the nodes in the sub-canvas, you need to verify whether there is a sub-canvas.
    if (
      isNodeInSubCanvas(node) &&
      hasSubCanvasNodes(this.workflowDocument, workflow.nodes)
    ) {
      Toast.warning(
        I18n.t(
          'workflow_encapsulate_toast_batch_or_loop',
          undefined,
          '在循环/批处理中解散的工作流不能包含循环或批处理节点',
        ),
      );
      return;
    }

    this.historyService.startTransaction();

    // Node-centric expansion
    this.encapsulateNodesService.decapsulateLayout(node, workflow.nodes);

    // Create Node
    const { idsMap, startNode, endNode, middleNodes } =
      await this.encapsulateNodesService.createDecapsulateNodes(
        node,
        workflow.nodes,
        node.parent?.id,
      );

    // Update variable reference after unblocking
    this.encapsulateVariableService.updateVarsAfterDecapsulate(node, {
      idsMap,
      startNode,
      endNode,
      middleNodes,
    });

    // Create a connection
    this.encapsulateLinesService.createDecapsulateLines({
      node,
      workflowJSON: workflow as WorkflowJSON,
      startNodeId: startNode.id,
      endNodeId: endNode.id,
      idsMap,
    });

    await delay(30);

    // Remove old node
    await this.encapsulateNodesService.deleteNodes([node]);
    this.historyService.endTransaction();
  }

  dispose() {
    this.onEncapsulateEmitter.dispose();
  }

  private async encapsulateNodes(
    name: string,
    nodes: WorkflowNodeEntity[],
  ): Promise<EncapsulateResult> {
    const ports =
      this.encapsulateLinesService.getValidEncapsulateConnectPorts(nodes);

    const encapsulateVars =
      this.encapsulateVariableService.getEncapsulateVars(nodes);

    const startEndRects =
      this.encapsulateNodesService.getEncapsulateStartEndRects(nodes);

    // Generate JSON and update the variable references in JSON
    const json = await this.encapsulateGenerateService.generateWorkflowJSON(
      nodes,
      {
        startEndRects,
      },
    );

    let workflowId;
    let flowMode = WorkflowMode.Workflow;
    if (
      this.encapsulateContext.isChatFlow &&
      encapsulateVars.startVars.USER_INPUT &&
      encapsulateVars.startVars.CONVERSATION_NAME
    ) {
      flowMode = WorkflowMode.ChatFlow;
    }

    try {
      // save process
      const res = await this.encapsulateApiService.encapsulateWorkflow({
        name,
        desc: name,
        json,
        flowMode,
      });
      workflowId = res?.workflowId;
    } catch (error) {
      logger.error(error);
      return this.encapsulateError('encapsulate workflow failed');
    }

    if (!workflowId) {
      return this.encapsulateError(
        'encapsulate workflow failed no workflowId returned',
      );
    }

    this.historyService.startTransaction();
    // Replace with calling process node
    const subFlowNode =
      await this.encapsulateNodesService.createEncapsulateNode(
        workflowId,
        name,
        nodes,
      );

    // Update variable references (upstream and downstream + encapsulate node itself input)
    this.encapsulateVariableService.updateVarsAfterEncapsulate(
      subFlowNode,
      nodes,
      encapsulateVars,
    );

    // Remove old node
    await this.encapsulateNodesService.deleteNodes(nodes);

    // Create a new connection
    const { inputLines, outputLines } =
      this.encapsulateLinesService.createEncapsulateLines(ports, subFlowNode);
    await delay(30);

    this.historyService.endTransaction();
    return {
      success: true,
      subFlowNode,
      inputLines,
      outputLines,
      workflowId,
      projectId: this.encapsulateContext.projectId,
    };
  }

  private encapsulateError(message: string): EncapsulateErrorResult {
    return {
      success: false,
      message,
    };
  }
}
