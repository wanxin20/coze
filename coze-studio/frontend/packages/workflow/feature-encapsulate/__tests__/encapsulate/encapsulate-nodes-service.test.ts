/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, beforeEach, it, expect } from 'vitest';
import { FlowNodeTransformData } from '@flowgram-adapter/free-layout-editor';
import {
  WorkflowDocument,
  type WorkflowNodeEntity,
} from '@flowgram-adapter/free-layout-editor';

import { complexMock } from '../workflow.mock';
import { createContainer } from '../create-container';
import { EncapsulateNodesService } from '../../src/encapsulate';

describe('encapsulate-nodes-service', () => {
  let workflowDocument: WorkflowDocument;
  let encapsulateNodesService: EncapsulateNodesService;

  beforeEach(() => {
    const container = createContainer();
    workflowDocument = container.get<WorkflowDocument>(WorkflowDocument);
    encapsulateNodesService = container.get<EncapsulateNodesService>(
      EncapsulateNodesService,
    );
  });

  it('should getNodesMiddlePoint', async () => {
    await workflowDocument.fromJSON(complexMock);

    const nodes = ['109408', '156471'].map(id =>
      workflowDocument.getNode(id),
    ) as WorkflowNodeEntity[];
    const point = encapsulateNodesService.getNodesMiddlePoint(nodes);

    expect(point).toEqual({
      x: 993.4484760125105,
      y: 489.11299357749374,
    });
  });

  it('should getNodesMiddlePoint by json', () => {
    const point = encapsulateNodesService.getNodesMiddlePoint(
      complexMock.nodes,
    );

    expect(point).toEqual({
      x: 808.8382078287623,
      y: 207.51796739960494,
    });
  });

  it('should createDecapsulateNode', async () => {
    await workflowDocument.fromJSON(complexMock);
    const json = {
      id: '1',
      type: 'test',
      meta: {
        position: {
          x: 0,
          y: 0,
        },
      },
      blocks: [
        {
          id: '2',
          type: 'test',
          meta: {
            position: {
              x: 5,
              y: 6,
            },
          },
        },
      ],
    };

    const idsMap = new Map<string, string>();
    const node = await encapsulateNodesService.createDecapsulateNode(
      json,
      {
        x: 170,
        y: 16.700000000000017,
      },
      idsMap,
    );
    expect(node.id).not.toBe('1');
    expect(node.flowNodeType).toBe(json.type);
    const transformData = node.getData(FlowNodeTransformData);
    expect(transformData.position).toEqual({ x: 170, y: 16.700000000000017 });

    const child = node.collapsedChildren[0];
    expect(child.id).not.toBe('2');
    expect(child.flowNodeType).toBe(json.blocks[0].type);
    const childTransformData = child.getData(FlowNodeTransformData);
    expect(childTransformData.position).toEqual({
      x: 175,
      y: 22.700000000000017,
    });

    expect(idsMap.get('1')).toBe(node.id);
    expect(idsMap.get('2')).toBe(child.id);
  });

  it('should decapsulateLayout', async () => {
    await workflowDocument.fromJSON(complexMock);
    await encapsulateNodesService.decapsulateLayout(
      workflowDocument.getNode('100001') as WorkflowNodeEntity,
      [complexMock.nodes[1], complexMock.nodes[2], complexMock.nodes[3]],
    );
    const json = await workflowDocument.toJSON();
    expect(json).toMatchSnapshot();
  });
});
