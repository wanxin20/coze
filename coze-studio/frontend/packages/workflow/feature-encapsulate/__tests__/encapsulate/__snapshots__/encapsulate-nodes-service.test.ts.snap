// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encapsulate-nodes-service > should decapsulateLayout 1`] = `
{
  "edges": [
    {
      "sourceNodeID": "100001",
      "targetNodeID": "154702",
    },
    {
      "sourceNodeID": "102906",
      "targetNodeID": "900001",
    },
    {
      "sourceNodeID": "177547",
      "targetNodeID": "154702",
    },
    {
      "sourceNodeID": "154702",
      "targetNodeID": "102906",
    },
    {
      "sourceNodeID": "177547",
      "targetNodeID": "109408",
    },
  ],
  "nodes": [
    {
      "data": undefined,
      "id": "100001",
      "meta": {
        "position": {
          "x": 180,
          "y": 26.700000000000017,
        },
      },
      "type": "1",
    },
    {
      "data": undefined,
      "id": "900001",
      "meta": {
        "position": {
          "x": 1743.609729023942,
          "y": 176.16991217034956,
        },
      },
      "type": "2",
    },
    {
      "data": undefined,
      "id": "154702",
      "meta": {
        "position": {
          "x": 987.7405729256998,
          "y": 245.06502111375198,
        },
      },
      "type": "3",
    },
    {
      "data": {
        "inputs": {
          "spaceId": "test_space_id",
          "workflowId": "test_workflow_id",
        },
      },
      "id": "102906",
      "meta": {
        "position": {
          "x": 848.7417419605051,
          "y": -297.0809682834506,
        },
      },
      "type": "9",
    },
    {
      "data": undefined,
      "id": "177547",
      "meta": {
        "position": {
          "x": -125.93331336641754,
          "y": 357.10168132837816,
        },
      },
      "type": "3",
    },
    {
      "data": undefined,
      "id": "109408",
      "meta": {
        "position": {
          "x": 1141.6692827086551,
          "y": 537.1820787709282,
        },
      },
      "type": "3",
    },
    {
      "data": undefined,
      "id": "156471",
      "meta": {
        "position": {
          "x": 984.2265002815602,
          "y": 712.1169030826604,
        },
      },
      "type": "3",
    },
  ],
}
`;
